File: backend/app/services/account_health_rollup.py
"""Project -> Account rollup for RAG Status notes: surfaces a project's own
health items so an Account Manager can pull individual ones into the
account's own register. Mirrors services/account_rollup.py, minus the Key
Metrics summing step — HealthDeclaration has no Draft/Submitted status and
no numeric fields to sum, so every ProjectHealthItem for the account's
projects at the given period is a candidate, with no report-status filter.
"""

from uuid import UUID

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.crud.account_health_declarations import account_health_item_crud
from app.models.health_declarations import ProjectHealthItem
from app.models.projects import Project
from app.schemas.account_health_declarations import AccountHealthItemCreate
from app.schemas.account_health_rollup import AccountHealthRollupItem, AccountHealthRollupResponse
from app.schemas.enums import RollupStatus
from app.services.account_rollup import RollupItemAlreadyHandledError, RollupItemNotFoundError


async def compute_account_health_rollup(
    db: AsyncSession, account_id: UUID, period_id: UUID
) -> AccountHealthRollupResponse:
    projects = (await db.execute(select(Project).where(Project.account_id == account_id))).scalars().all()
    if not projects:
        return AccountHealthRollupResponse(items=[])

    project_by_id = {p.id: p for p in projects}
    health_items = (
        (
            await db.execute(
                select(ProjectHealthItem).where(
                    ProjectHealthItem.project_id.in_(project_by_id.keys()),
                    ProjectHealthItem.period_id == period_id,
                )
            )
        )
        .scalars()
        .all()
    )

    items = [
        AccountHealthRollupItem(
            id=item.id,
            project_id=item.project_id,
            project_code=project_by_id[item.project_id].project_code,
            project_name=project_by_id[item.project_id].project_name,
            category=item.category,
            description=item.description,
            account_rollup_status=item.account_rollup_status,
            rolled_up_account_item_id=item.rolled_up_account_item_id,
        )
        for item in health_items
    ]

    return AccountHealthRollupResponse(items=items)


async def pull_health_rollup_item(db: AsyncSession, account_id: UUID, project_item_id: UUID):
    item = await db.get(ProjectHealthItem, project_item_id)
    if item is None:
        raise RollupItemNotFoundError
    project = await db.get(Project, item.project_id)
    if project is None or project.account_id != account_id:
        raise RollupItemNotFoundError
    if item.account_rollup_status != RollupStatus.PENDING:
        raise RollupItemAlreadyHandledError

    account_item = await account_health_item_crud.create(
        db,
        AccountHealthItemCreate(
            period_id=item.period_id,
            category=item.category,
            description=f"{project.project_code} — {item.description}",
        ),
        account_id=account_id,
    )

    item.account_rollup_status = RollupStatus.PULLED
    item.rolled_up_account_item_id = account_item.id
    await db.flush()

    return account_item

File: frontend/src/lib/api/account-health-rollup.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { api } from "./client";
import type { AccountHealthItem, HealthCategory } from "./account-health-declarations";

export type RollupStatus = "Pending" | "Pulled" | "Ignored";

export type AccountHealthRollupItem = {
  id: string;
  project_id: string;
  project_code: string;
  project_name: string;
  category: HealthCategory;
  description: string;
  account_rollup_status: RollupStatus;
  rolled_up_account_item_id: string | null;
};

export type AccountHealthRollupResponse = {
  items: AccountHealthRollupItem[];
};

// Project -> Account rollup for RAG Status notes: surfaces the account's
// projects' own health items so they can be pulled into (or dismissed from)
// the account's own register. Mirrors account-rollup.ts, minus Key Metrics
// (RAG Status has no numeric fields to sum).
export function useAccountHealthRollup(accountId: string | null, periodId: string | null) {
  return useQuery({
    queryKey: ["account-health-rollup", accountId, periodId],
    queryFn: () => api.get<AccountHealthRollupResponse>(`/accounts/${accountId}/health-rollup?period_id=${periodId}`),
    enabled: !!accountId && !!periodId,
  });
}

function invalidateHealthRollup(queryClient: ReturnType<typeof useQueryClient>, accountId: string | null) {
  queryClient.invalidateQueries({ queryKey: ["account-health-rollup", accountId] });
}

export function usePullHealthRollupItem(accountId: string | null) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (projectItemId: string) =>
      api.post<AccountHealthItem>(`/accounts/${accountId}/health-rollup/pull`, { project_item_id: projectItemId }),
    onSuccess: (item) => {
      invalidateHealthRollup(queryClient, accountId);
      queryClient.invalidateQueries({
        queryKey: ["account-health-items", accountId, item.period_id, item.category],
      });
    },
  });
}

// Items on one account's rollup panel come from several different
// projects, so the project id travels with each call rather than being
// fixed at the hook level.
export function useSetHealthItemRollupStatus(accountId: string | null) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({
      projectId,
      itemId,
      status,
    }: {
      projectId: string;
      itemId: string;
      status: "Pending" | "Ignored";
    }) => api.patch(`/projects/${projectId}/health-items/${itemId}/rollup-status`, { status }),
    onSuccess: () => invalidateHealthRollup(queryClient, accountId),
  });
}
